stablehlo/dialect/TypeInference.cpp

```diff
@@ -1141,8 +1136,14 @@ LogicalResult inferSliceOp(Optional<Location> location, Value operand,
       return emitOptionalError(location, "stride must be positive but got ",
                                strideVals[i], " in dimension ", i);
 
-    shape.push_back(inferSliceDim(rankedTy.getDimSize(i), start[i], limit[i],
-                                  strideVals[i]));
+    if (!hlo::isDynamicDimSize(rankedTy.getDimSize(i))) {  // static dim
+      shape.push_back(inferSliceDim(rankedTy.getDimSize(i), start[i], limit[i],
+                                    strideVals[i]));
+      if (!inputBounds.empty())
+        resultBounds.push_back(ShapedType::kDynamicSize);
+    } else  // dynamic dim with bounds
+      resultBounds.push_back(
+          inferSliceDim(inputBounds[i], start[i], limit[i], strideVals[i]));
```

Comment:
inferSliceDim is used only once, so let's inline it into the callsite, so that readers don't have to jump back and forth between the definition and the callsite. Then we can combine both static sizes and dynamic bounds in a single snippet of code in a very straightforward way:

```C++
// Assuming that `bool isDynamicDim = hlo::isDynamicDimSize(inputShape[i]);` is declared above.
int64_t operandSizeOrBound = isDynamicDim ? inputBounds[i] : rankedTy.getDimSize(i);
int64_t resultSizeOrBound = static_cast<int64_t>(llvm::divideCeil(limit[i] - start[i], strideVals[i]));
(isDynamicDim ? resultBounds : shape).push_back(resultSizeOrBound);
```

Reply:
Thanks for the demo! Your suggested is used except the last line "(isDynamicDim ? resultBounds : shape)", we have 3 cases instead of 2:
1. unbounded dynamic
2. bounded dynamic
3. static
So a "?:" is not quite useful here. So I propose my new code in new commit. Please let me know what you think of it.

Reply:
Here's the full code snippet that I had in mind. I liked your original idea to `continue` early for the unbounded dynamic case and based the proposal on that:

```C++
  for (int64_t i = 0, e = rank; i != e; i++) {
    // P3.
    if (start[i] < 0)
      return emitOptionalError(location, "negative start index ", start[i],
                               " in dimension ", i);
    // P5.
    if (start[i] > limit[i])
      return emitOptionalError(location, "start index ", start[i],
                               " is larger than limit index ", limit[i],
                               " in dimension ", i);
    // P6.
    if (strideVals[i] <= 0)
      return emitOptionalError(location, "stride must be positive but got ",
                               strideVals[i], " in dimension ", i);

    bool isDynamicDim = hlo::isDynamicDimSize(inputShape[i]);
    if (isDynamicDim) {
      shape.push_back(ShapedType::kDynamicSize);
      if (inputBounds.empty()) continue;
    }

    int64_t operandSizeOrBound = isDynamicDim ? inputBounds[i] : rankedTy.getDimSize(i);
    StringRef sizeOrBound = isDynamicDim ? "bound" : "size";

    // P4.
    if (limit[i] > operandSizeOrBound)
      return emitOptionalError(location, "limit index ", limit[i],
                               " is larger than dimension ", sizeOrBound,
                               rankedTy.getDimSize(i), " in dimension ", i);

    int64_t resultSizeOrBound = static_cast<int64_t>(
        llvm::divideCeil(limit[i] - start[i], strideVals[i]));
    (isDynamicDim ? resultBounds : shape).push_back(resultSizeOrBound);
  }
```

What I like about this proposal is that it allows to write P4 and the final computation in an almost generic way - only 1 conditional is needed. For comparison, with the current PR, we need 3 `isDynamicDimSize` checks and 3 `!inputBounds.empty` checks.

Reply:
Thanks a lot for the code snippet above! and used with 2 changes:
1. in P4, `limit[i] > operandSizeOrBound` -> `operandSizeOrBound != ShapedType::kDynamicSize &&
        limit[i] > operandSizeOrBound`. Because in bounded dynamic case, the `inputBounds[i]` (`operandSizeOrBound`) could be -1 even for a dynamic dim. We should skip P4 check in such case.
2. in the end, I add 2 new lines: 
```
if (!isDynamicDim && !inputBounds.empty())  // use bound=-1 for static dim
  resultBounds.push_back(ShapedType::kDynamicSize);
``` 
for the static dim when dealing with bounded dynamism.

For (2), an alternative way is to initializing the `resultBounds`/`shape` with (size_k, -1). But I do not see it worth an big change like that.

Reply:
You're right, there's a possibility which I missed in the code snippet in my previous comment - fully dynamic dimension (? for size, ? for bound). 

This prompted me to write down all the possibilities that can happen, which can be found in the table below, given that `Z = llvm::divideCeil(limit - start, stride)`, and I think I found some bugs in the current implementation:

|    | Input Shape | Input Bound | Action             | Output Shape | Output Bound |
|----|-------------|-------------|--------------------|--------------|--------------|
| C1 | X           | N/A         | Check `X <= limit` | Z            | N/A          |
| C2 | X           | ?           | Check `X <= limit` | Z            | N/A          |
| C3 | X           | Y           | Impossible         | Impossible   | Impossible   |
| C4 | ?           | N/A         | No checks          | Z            | N/A          |
| C5 | ?           | ?           | No checks          | Z            | N/A          |
| C6 | ?           | Y           | Check `Y <= limit` | Z            | N/A          |

I think that the only thing that changes if we vary different possibilities between input shapes and input bounds is P4 - if either the dim size or the dim bound are static, then we need to check whether they are `<= limit`. But the output should always be the same - fully static shape without bounds (because `start`, `limit` and `strides` unambiguously specify a static dim size):

```C++
  ArrayRef<int64_t> inputBounds = encodingToBounds(rankedTy.getEncoding());
  SmallVector<int64_t> shape(rank, ShapedType::kDynamicSize);
  SmallVector<int64_t> resultBounds(inputBounds.size(),
                                    ShapedType::kDynamicSize);

  for (int64_t i = 0, e = rank; i != e; i++) {
    // P3.
    if (start[i] < 0)
      return emitOptionalError(location, "negative start index ", start[i],
                               " in dimension ", i);

    // P4.
    bool isStaticDim = !hlo::isDynamicDimSize(rankedTy.getDimSize(i));
    bool isStaticBound =
        !inputBounds.empty() && !hlo::isDynamicDimSize(inputBounds[i]);
    if (isStaticDim || isStaticBound) {
      int64_t operandSizeOrBound =
          isStaticDim ? rankedTy.getDimSize(i) : inputBounds[i];
      StringRef sizeOrBound = isStaticDim ? "size" : "bound";
      if (limit[i] > operandSizeOrBound)
        return emitOptionalError(location, "limit index ", limit[i],
                                 " is larger than dimension ", sizeOrBound, " ",
                                 operandSizeOrBound, " in dimension ", i);
    }

    // P5.
    if (start[i] > limit[i])
      return emitOptionalError(location, "start index ", start[i],
                               " is larger than limit index ", limit[i],
                               " in dimension ", i);
    // P6.
    if (strideVals[i] <= 0)
      return emitOptionalError(location, "stride must be positive but got ",
                               strideVals[i], " in dimension ", i);

    shape[i] = static_cast<int64_t>(
        llvm::divideCeil(limit[i] - start[i], strideVals[i]));
  }
```

I've created a pull request to your branch that updates the logic and the tests accordingly: https://github.com/zhouxin913/stablehlo/pull/1.

Reply:
The table makes it extremely clear! and I agree the big conclusions we have here are:
1)  The new logic we summarize in the table will not affect the old users in any way (who do not use bounds at all) 
2) for some ops/cases, the bounds in input will lead to bounds in output. But for some other ops like this SliceOp, it may replace the bounds with static dims. This will be expected to be case by case for every op in the future. 

https://github.com/zhouxin913/stablehlo/pull/1 is approved and merged, which update this PR with a new commit.